### PR TITLE
feat: update injector away from rouille to axum

### LINKED ISSUE #592: Zarf-Seed-Registry Installation Fails on Init with Deployment is not ready: zarf/zarf-docker-registry error
### Environment
Device and OS: Azure AKS Linux Ubuntu 20.04
App version: 0.19.6
Kubernetes distro being used: AKS Kubernetes V 1.22.6
Other:

### Steps to reproduce
1. Create an AKS Cluster
2. Run `zarf init --components git-server`.

### Expected result
Command succeeds and Zarf is initialized in the cluster.

### Actual Result
```
The following message repeats until the init run timesout
Deployment is not ready: zarf/zarf-docker-registry. 0 out of 1 expected pods are ready
```

output of `kubectl -n zarf get events`

```
LAST SEEN   TYPE      REASON              OBJECT                                       MESSAGE
15m         Normal    Scheduled           pod/injector                                 Successfully assigned zarf/injector to aks-agentpool-40722291-vmss000001
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-018" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-023" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-013" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-008" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-009" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-019" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "stage1" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-027" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 MountVolume.SetUp failed for volume "zarf-payload-007" : failed to sync configmap cache: timed out waiting for the condition
15m         Warning   FailedMount         pod/injector                                 (combined from similar events): MountVolume.SetUp failed for volume "zarf-payload-027" : failed to sync configmap cache: timed out waiting for the condition
15m         Normal    Scheduled           pod/zarf-docker-registry-789d8ddfb8-4pfgj    Successfully assigned zarf/zarf-docker-registry-789d8ddfb8-4pfgj to aks-agentpool-40722291-vmss000000
14m         Normal    Pulling             pod/zarf-docker-registry-789d8ddfb8-4pfgj    Pulling image "127.0.0.1:32178/library/registry:2.7.1"
14m         Warning   Failed              pod/zarf-docker-registry-789d8ddfb8-4pfgj    Failed to pull image "127.0.0.1:32178/library/registry:2.7.1": rpc error: code = Unknown desc = failed to pull and unpack image "127.0.0.1:32178/library/registry:2.7.1": failed to resolve reference "127.0.0.1:32178/library/registry:2.7.1": failed to do request: Head "https://127.0.0.1:32178/v2/library/registry/manifests/2.7.1": http: server gave HTTP response to HTTPS client
14m         Warning   Failed              pod/zarf-docker-registry-789d8ddfb8-4pfgj    Error: ErrImagePull
34s         Normal    BackOff             pod/zarf-docker-registry-789d8ddfb8-4pfgj    Back-off pulling image "127.0.0.1:32178/library/registry:2.7.1"
13m         Warning   Failed              pod/zarf-docker-registry-789d8ddfb8-4pfgj    Error: ImagePullBackOff
15m         Normal    SuccessfulCreate    replicaset/zarf-docker-registry-789d8ddfb8   Created pod: zarf-docker-registry-789d8ddfb8-4pfgj
15m         Normal    ScalingReplicaSet   deployment/zarf-docker-registry              Scaled up replica set zarf-docker-registry-789d8ddfb8 to 1
```

output of `kubectl -n zarf get all`
```
NAME                                        READY   STATUS             RESTARTS   AGE
pod/injector                                1/1     Running            0          40m
pod/zarf-docker-registry-789d8ddfb8-4pfgj   0/1     ImagePullBackOff   0          40m

NAME                           TYPE       CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
service/zarf-docker-registry   NodePort   10.0.16.59     <none>        5000:31999/TCP   40m
service/zarf-injector          NodePort   10.0.144.122   <none>        5000:32178/TCP   40m

NAME                                   READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/zarf-docker-registry   0/1     1            0           40m

NAME                                              DESIRED   CURRENT   READY   AGE
replicaset.apps/zarf-docker-registry-789d8ddfb8   1         1         0       40m
```

### Severity/Priority
😕  Blocked on deploying zarf packages to Azure AKS

### PR DESCRIPTION
**Pull request by schristoff**
Fixes #592

### CODE REVIEW

File: src/injector/src/main.rs [hunk 1]
```diff
@@ -180,46 +159,74 @@ fn handle_get_manifest(root: &Path, name: &String, tag: &String) -> Response {
...
+#[tokio::main]
+async fn main() {
     let args: Vec<String> = env::args().collect();
 
-    match args.len() {
-        2 => {
-            let payload_sha = &args[1];
-            unpack(payload_sha);
+    println!("unpacking: {}", args[1]);
+    let payload_sha = &args[1];
```
Comment (schristoff) [inline src/injector/src/main.rs:219]: Double check if this is okay - unsure if we need to take in two

File: packages/zarf-registry/armregistry-zarf
Comment (eddiezane) [inline packages/zarf-registry/armregistry-zarf:1]: Need to remove this and the other binary from when you were deving.

File: src/injector/README.md [hunk 2]
```diff
@@ -31,3 +56,41 @@ echo "Linux binary size: $size_linux"
...
+
+    files:
+      # Rust Injector Binary
+      - source: ../../src/injector/target/aarch64-unknown-linux-musl/release/zarf-injector
+        target: "###ZARF_TEMP###/zarf-injector"
+        <!-- shasum: "###ZARF_PKG_TMPL_INJECTOR_ARM64_SHASUM###" -->
+        executable: true
+```
+
+In Zarf Root Directory, run `make init-package`, if you are running on an Apple Silicon
```
Comment (eddiezane) [inline src/injector/README.md:81]: Do we need a step to run make from the injector src or is that elsewhere?

File: src/injector/src/main.rs [hunk 3]
```diff
@@ -86,90 +95,82 @@ fn unpack(sha_sum: &String) {
...
+    .route("/v2/", get(|| async { 
+    Response::builder()
+        .status(StatusCode::OK)
+        .header("Content-Type", "application/json; charset=utf-8")
+        .header("Docker-Distribution-Api-Version", "registry/2.0")
+        .header("X-Content-Type-Options", "nosniff")
+        .body(Body::empty())
+        .unwrap()
+    }))
+    .route("/v2", get(|| async {
```
Comment (eddiezane) [inline src/injector/src/main.rs:121]: I think axum will route these to the same place so we can probably drop the second?

https://github.com/tokio-rs/axum/issues/1118

File: src/injector/src/main.rs [hunk 4]
```diff
@@ -86,90 +95,82 @@ fn unpack(sha_sum: &String) {
...
-            return handle_get_digest(&root, &digest);
-        }
-    }
+async fn handler(Path(path): Path<String>) -> Response {
+    println!("request: {}", path);
+    let path = &path;
+    let manifest = Regex::new("(.+)/manifests/(.+)").unwrap();
+    let blob = Regex::new(".+/([^/]+)").unwrap();
+
+    if manifest.is_match(path){
```
Comment (eddiezane) [inline src/injector/src/main.rs:138]: Running `cargo fmt` should clean up some of these style things.

File: src/injector/src/main.rs [hunk 5]
```diff
@@ -86,90 +95,82 @@ fn unpack(sha_sum: &String) {
...
+        .into_response()
+  }
 }
 
 /// Handles the GET request for the manifest (only returns a OCI manifest regardless of Accept header)
-fn handle_get_manifest(root: &Path, name: &String, tag: &String) -> Response {
-    let index = fs::read_to_string(root.join("index.json")).expect("read index.json");
+async fn handle_get_manifest(name: String, reference: String) -> Response {
+    println!("name {}, reference {}", name, reference);
+    let index = fs::read_to_string(PathBuf::from("/zarf-seed").join("index.json")).expect("read index.json");
```
Comment (eddiezane) [inline src/injector/src/main.rs:161]: nit: style for `expect` https://doc.rust-lang.org/std/option/enum.Option.html#recommended-message-style

File: src/injector/src/main.rs [hunk 6]
```diff
@@ -180,46 +181,82 @@ fn handle_get_manifest(root: &Path, name: &String, tag: &String) -> Response {
             }
         }
     }
+    if !sha_manifest.is_empty() {
```
Comment (eddiezane) [inline src/injector/src/main.rs:184]: nit: consider inverting this statement to drop the `!`.

**Review by eddiezane — CHANGES_REQUESTED** [review]
Nice job on this!

File: src/injector/README.md (hunk 2 above)
Comment (schristoff) [reply src/injector/README.md:91]: updated this to use make cmds

**Review by lucasrod16 — COMMENTED** [review]
It's working on my machine 🎉 nice work!

Could we go over the Rust changes in one of our syncs this week? I'm not too familiar with Rust and it would help me understand them better.

File: src/injector/README.md [hunk 7]
```diff
@@ -17,6 +18,30 @@ rustup target add x86_64-unknown-linux-musl
...
+rustup toolchain install --force-non-host stable-x86_64-unknown-linux-musl
+```
+```
+cross build --target x86_64-unknown-linux-musl --release
+
+cross build --target aarch64-unknown-linux-musl --release
+```
+
+This will build into `target/*--unknown-linux-musl/release`
+
```
[the remark this answers is not in the document]
Comment (schristoff-du) [reply src/injector/README.md:42]: I will add this to the readme as a different approach. I had issues with running this, but I think it's just maybe a my machine thing.

File: .github/workflows/build-rust-injector.yml [hunk 8]
```diff
@@ -27,8 +27,9 @@ jobs:
 
       - name: "Build Rust Binary for x86_64 and arm64"
         run: |
-          make build-injector-linux
-          cd src/injector/target
+          cd src/injector
+          make cross-injector-linux
```
Comment (eddiezane) [inline .github/workflows/build-rust-injector.yml:31]: Saw @lucasrod16's comments about cross below. We should probably consciously decide either yes or no cross for the one we publish.

File: src/injector/src/main.rs [hunk 9]
```diff
@@ -93,83 +101,77 @@ fn unpack(sha_sum: &String) {
...
+    let blob = Regex::new(".+/([^/]+)").unwrap();
+
+    if manifest.is_match(path) {
+        let caps = manifest.captures(path).unwrap();
+        let name = caps.get(1).unwrap().as_str().to_string();
+        let reference = caps.get(2).unwrap().as_str().to_string();
+        handle_get_manifest(name, reference).await
+    } else if blob.is_match(&path) {
+        let caps = blob.captures(path).unwrap();
+        let tag = caps.get(1).unwrap().as_str().to_string();
```
Comment (eddiezane) [inline src/injector/src/main.rs:148]: Probably not blocking or even on this PR but all these `unwrap`'s should eventually be handled properly.

https://docs.rs/axum/latest/x86_64-pc-windows-msvc/axum/error_handling/index.html#axums-error-handling-model

**Review by eddiezane — COMMENTED** [review]
LGTM! Left a few extra comments

File: .github/workflows/build-rust-injector.yml (hunk 8 above)
Comment (schristoff) [reply .github/workflows/build-rust-injector.yml:31]: His comment is about the Readme, we should not be running docker containers in GHAs to build rust (rust comes on gha by default) but we can make it so it uses cargo instead.

File: .github/actions/k3d/action.yaml [hunk 10]
```diff
@@ -7,5 +7,5 @@ runs:
     - run: "curl -s https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash"
       shell: bash
 
-    - run: k3d cluster delete && k3d cluster create --k3s-arg="--disable=traefik@server:0" --image="rancher/k3s:v1.28.4-k3s2"
```
[the remark this answers is not in the document]
Comment (schristoff) [reply .github/actions/k3d/action.yaml:10]: It was tagged to an image that would make it pass CI, now with these changes that can be removed.

File: src/injector/src/main.rs (hunk 9 above)
Comment (schristoff-du) [reply src/injector/src/main.rs:148]: I'm not opposed to it, [but it may put us over size limit](https://crates.io/crates/error_handling)

**Review by lucasrod16 — APPROVED** [review]
lgtm

**Review by eddiezane — DISMISSED** [review]

**Review by lucasrod16 — APPROVED** [review]

**Pull request merged**